### PR TITLE
Add testcases beamtruss

### PR DESCRIPTION
**Pull request by KlausBSautter**
Hi ,
I added test cases for the truss and the beam element. There were none yet and I think it's important to have them. Both test files should also accord to the new way of writing small tests.

Additionally I fixed a few things to make it possible to use the feast eigen-wert solver with the beam and the new static eigen-wert solver of @peterjwilson.

Klaus

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp]: Would you please use the new [header template](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/templates/header_template)?

**Comment by pooyan-dadvand** [comment]
For some reason I cannot see the diffs correctly.....

**Review by pooyan-dadvand — COMMENTED** [review]
I see that there are a lot of Matrix creation in the calculation. Many of them can be avoided and some can be changed to bounded matrix. Creation of the Matrix is really expensive.

**Comment by pooyan-dadvand** [comment]
Adding `?w=1` to the url of file diffs takes out white space difference and then it is more clear!

**Comment by KlausBSautter** [comment]
@pooyan-dadvand  thanks for looking at it. I will have a look myself. Where exactly are you referring to when talking about the matrix creation?
Klaus

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Get it a s a Vector and then access the components, so to minimize database access

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp]: Use the standard header

**Review by RiccardoRossi — APPROVED** [review]
Hi, 

I just put a performance comment, and one about using the standard header.

Just change that and merge

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Can be a BoundedMatrix? You know the size.

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Returning the Matrix by value may be a performance issue as I'm not sure if this matrix implements move semantic.

**Comment by pooyan-dadvand** [comment]
OK, I will mention some of them.

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Same about the geometry

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: is it a 2 node beam or can be other geometry? if is only 2 please put 2 as number of node so code can be optimized more.

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Again, better to avoid returning matrix by value. you can pass it as an argument.

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: If you know the local_size and size in compile time it would be better to use bounded matrix

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: If you know the size please use array_1d

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: If you know the size please use array_1d

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: array_1d<double, 6>

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: bounded matrix 3x3?

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: bounded matrix 3x3?

Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: bounded matrix 12x6?

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I have tried to mention you some of the Matrix and vector that possibly can be improved to fix size one. Nevertheless there are several of them and you may find the rest by yourself. In general any auxiliary one is better to be avoided.

**Comment by pooyan-dadvand** [comment]
@KlausBSautter thanks for your changes, however I think I have mislead you in using known numbers.
My suggestion was to put:

```c++
  constexpr int number_of_nodes =2;
  constexpr int dimension = 3;
  constexpr unsigned int local_size = number_of_nodes * dimension;
```

and then use them as constexpr variable to build `array_1d` and `bounded_matrix`

Note that using numbers like 6, 2, ... in a code is not really good because they are not descriptive as variable name and understanding the code becomes harder

**Comment by KlausBSautter** [comment]
@pooyan-dadvand  I understand. So if I am passing a

**Comment by KlausBSautter** [comment]
@pooyan-dadvand Hi I also adjusted the beam element and I reversed some of the changes for the truss element. If I follow your advice to avoid explicitly writing int numbers I cannot use bounded matrices / vectors as function parameters without explicitly writing an int value. 
Please correct me if I am wrong

**Comment by pooyan-dadvand** [comment]
The trick is using `constexpr`as I described in my example:

```c++
  constexpr int number_of_nodes =2;
  constexpr int dimension = 3;
  constexpr unsigned int local_size = number_of_nodes * dimension;
```
without `constexpr` it won't compile

**Comment by KlausBSautter** [comment]
As global variables in the header file?

**Comment by pooyan-dadvand** [comment]
No, as local variable in the method or if you want as member variable of class.

**Comment by philbucher** [comment]
@KlausBSautter do not merge this until #651 is merged! you included this branch in your pull request and would therefore silently merge it!

**Comment by KlausBSautter** [comment]
@philbucher  yes I have to do changes anyway, thanks

**Comment by KlausBSautter** [comment]
what about #651 ? Everything seems to be fine there and it is a necessary feature

**Comment by KlausBSautter** [comment]
I will close this request until #651 is merged to be sure

**Comment by KlausBSautter** [comment]
this is not ready to be merged. Ill reopen it as soon as it is

**Comment by KlausBSautter** [comment]
Alright. @pooyan-dadvand  I guess I changed everything, what do you think about it ?

**Comment by philbucher** [comment]
@KlausBSautter I fixed the issue with the rotations/reactions => you also had to add the rotation reaction dofs (TORQUE)
Also I used more core processes form the VariableUtils
TODO for you: Use the VariableUtils also in the truss test

**Comment by philbucher** [comment]
Almost forgot, most important, the tests take waaaay too long. From how I understood @RiccardoRossi they should run in a fraction of a sec.
On my (good!) laptop only the beam test takes ~1.5-2sec
This should be reduced, we can speak on Monday

**Comment by KlausBSautter** [comment]
@philbucher thanks for your help. I reduced the time to run them to under 1 sec each. Is this ok?

**Comment by KlausBSautter** [comment]
ok now the test patches take 0.033 sec together and the big tests are added. If everything is ok now we can merge. @pooyan-dadvand  @philbucher

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: I would suggest to use array_1d which is faster and is more standard in Kratos

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp
Comment (pooyan-dadvand) [inline applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp]: As they are static member variables they should named like: `msElementSize`, `msLocalSize` etc where "m" is for member and "s" is for static.

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
It is almost perfect! Nevertheless there are some very minor improvments:
- Changing the `bounded_vector` to `array_1d` would improve more the performance. They have the same interface, so a "replace all" would work for you.
- Correcting the styling for member variables would help the readability of the code.

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: as stated in issue #664 the usage of array_1d in combination with cross products needs an inverse input of the arrays. This is why I try to avoid using array_1d

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp]: alright sounds logical

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp
Comment (pooyan-dadvand) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Make sense!

Comment (pooyan-dadvand) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp]: Now I see, we should move with that issue. For now is ok what you did....

File: applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp
Comment (pooyan-dadvand) [reply applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.hpp]: So I will wait for this change

**Review by pooyan-dadvand — APPROVED** [review]
👍

**Comment by KlausBSautter** [comment]
@pooyan-dadvand I changed the names but I still like to wait until #850 from @philbucher  is merged so I can adjust my tests to the new changes.

**Comment by KlausBSautter** [comment]
I just merged this branch and changed my tests in #850

**Pull request merged**
